src/hotspot/share/opto/divnode.cpp

```diff
@@ -412,36 +378,110 @@ static Node *transform_long_divide( PhaseGVN *phase, Node *dividend, jlong divis
+    }
 
-      // Get a 0 or -1 from the sign of the dividend.
-      Node *addend0 = mul_hi;
-      Node *addend1 = phase->transform(new RShiftLNode(dividend, phase->intcon(N-1)));
+    // Get a 0 or -1 from the sign of the dividend.
+    Node *addend0 = mul_hi;
+    Node *addend1 = phase->transform(new RShiftLNode(dividend, phase->intcon(N-1)));
 
-      // If the divisor is negative, swap the order of the input addends;
-      // this has the effect of negating the quotient.
-      if (!d_pos) {
-        Node *temp = addend0; addend0 = addend1; addend1 = temp;
+    // If the divisor is negative, swap the order of the input addends;
+    // this has the effect of negating the quotient.
+    if (!d_pos) {
+      Node *temp = addend0; addend0 = addend1; addend1 = temp;
+    }
+
+    // Adjust the final quotient by subtracting -1 (adding 1)
+    // from the mul_hi.
+    q = new SubLNode(addend0, addend1);
+  }
+
+  return q;
+}
+
+//--------------------------transform_long_udivide-----------------------------
+// Convert an unsigned division by constant divisor into an alternate Ideal graph.
+// Return NULL if no transformation occurs.
+static Node* transform_long_udivide( PhaseGVN* phase, Node* dividend, julong divisor ) {
+  assert(divisor > 1, "invalid constant divisor");
+
+  // Result
+  Node* q = NULL;
+  if (is_power_of_2(divisor)) {
+    int l = log2i_exact(divisor);
+    q = new URShiftLNode(dividend, phase->intcon(l));
+  } else {
+    if (!Matcher::match_rule_supported(Op_UMulHiL)) {
+      return nullptr; // Don't bother
+    }
+
+    // Attempt the julong constant divide -> multiply transform found in
+    //   "Division by Invariant Integers using Multiplication"
+    //     by Granlund and Montgomery
+    // See also "Hacker's Delight", chapter 10 by Warren.
+
+    // Unsigned extension of dividend
+    jlong magic_const;
+    jint shift_const;
+    bool magic_const_ovf;
+    magic_long_unsigned_divide_constants(divisor, magic_const, shift_const, magic_const_ovf);
+    assert(shift_const >= 0 && shift_const < 65, "sanity");
+
+    Node* magic = phase->longcon(magic_const);
+    Node* mul_hi = phase->transform(new UMulHiLNode(dividend, magic));
+
+    if (!magic_const_ovf) {
+      assert(shift_const < 64, "sanity");
+      q = new URShiftLNode(mul_hi, phase->intcon(shift_const));
+    } else {
+      // If the multiplication does not overflow 128-bit unsigned int, we can
+      // do the addition right after
+      const TypeLong* dividend_type = phase->type(dividend)->is_long();
+      julong max_dividend;
+      if (dividend_type->_hi < 0 || dividend_type->_lo >= 0) {
+        max_dividend = julong(dividend_type->_hi);
+      } else {
+        max_dividend = max_julong;
       }
 
-      // Adjust the final quotient by subtracting -1 (adding 1)
-      // from the mul_hi.
-      q = new SubLNode(addend0, addend1);
+      // Just do the minimum for now
```

Comment:
I mean we can do better, we are trying to prove that the multiplication does not overflow a u128, since the constant is a u65 if the dividend is a u63 then the inequality holds, we can have a more strict bound since the constant is known already.

Reply:
I will add more clarification to this in the comment, basically the operation is `x * (M1 + 2**64) / 2**s` = `((x * M1) / 2**64 + 1) / 2**(s - 64)`.